File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
       List<List<Integer>> ans = new ArrayList<>();
        permut(0, nums, ans);
        return ans;
    }
    
    public static void permut(int index, int[] nums, List<List<Integer>> ans){
        if(index==nums.length){
            List<Integer> ds = new ArrayList<>();
            for(int i=0; i<nums.length; i++){
                ds.add(nums[i]);
            }
            ans.add(new ArrayList<>(ds));
            return;
        }
        for(int i=index; i<nums.length; i++){
            swap(i, index, nums);
            permut(index+1, nums, ans);
            swap(i, index, nums);
        }
    }
    
    public static void swap(int i, int index, int[] nums){
        int t = nums[i];
        nums[i]=nums[index];
        nums[index]=t;
    }
}

File: 0046-permutations/0046-permutations.py
from typing import List

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        used = set()

        def backtrack(temp):
            if len(temp) == len(nums):
                ans.append(temp[:])
                return

            for num in nums:
                if num in used:
                    continue

                temp.append(num)
                used.add(num)

                backtrack(temp)

                temp.pop()
                used.remove(num)

        backtrack([])
        return ans